### PR TITLE
Implement Webdriver-based Safari test harness

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This adds support for Webdriver-based browser tests and changes the macos Safari implementation to use this method. There is a fair amount of refactoring to the Browser and BrowserEnvironment classes and the test platform that uses them.

This is part of a fix for https://github.com/flutter/flutter/issues/99885. We still need to make some changes to the LUCI recipe repo in order to fix the rest.

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 112 new digest(s) on patchset 7.
View them at https://flutter-engine-gold.skia.org/cl/github/33757

**Comment by skia-gold** [comment]
Gold has detected about 218 new digest(s) on patchset 8.
View them at https://flutter-engine-gold.skia.org/cl/github/33757

**Comment by skia-gold** [comment]
Gold has detected about 87 new digest(s) on patchset 9.
View them at https://flutter-engine-gold.skia.org/cl/github/33757

File: lib/web_ui/dev/webdriver_browser.dart [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+  @override
+  Future<void> prepare() async {
+    _driverProcess = await spawnDriverProcess();
+
+    _driverProcess.stderr
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen((String error) {
+      print('[Webdriver] $error');
```
Comment (mdebbar) [inline lib/web_ui/dev/webdriver_browser.dart:29]: Maybe add `"[Error]"` so it's easier to distinguish errors from normal logs?
```suggestion
      print('[Webdriver][Error] $error');
```

File: lib/web_ui/dev/webdriver_browser.dart [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        return WebDriverBrowser(driver, url);
+      } on SocketException {
+        // Sometimes we may try to connect before the web driver port is ready. Retry.
+        print('Failed to connect to webdriver process. Retrying in 100 ms');
+        await Future<void>.delayed(const Duration(milliseconds: 100));
+      } catch (exception) {
+        print('Exception while creating webdriver: $exception');
+        rethrow;
+      }
+    }
```
Comment (mdebbar) [inline lib/web_ui/dev/webdriver_browser.dart:60]: This could become a real infinite loop (e.g. if webdriver never starts for whatever reason). Since there's no timeout or max retry logic here, are you relying on the test suite's timeout to stop this loop? If that's the case, it would be nice to clarify that in a comment.

**Review by mdebbar — APPROVED** [review]
Overall, looks good to me with some minor nits and comments.

**Comment by skia-gold** [comment]
Gold has detected about 148 new digest(s) on patchset 10.
View them at https://flutter-engine-gold.skia.org/cl/github/33757

File: lib/web_ui/dev/webdriver_browser.dart [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  @override
+  Future<void> cleanup() async {
+    _driverProcess.kill();
+  }
+
+  @override
+  Future<Browser> launchBrowserInstance(Uri url, {bool debug = false}) async {
+    for (;;) {
```
Comment (yjbanov) [inline lib/web_ui/dev/webdriver_browser.dart:47]: nit: I think `while (true)` is more idiomatic in Dart

File: lib/web_ui/dev/webdriver_browser.dart [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      try {
+        final WebDriver driver = await createDriver(
+          uri: driverUri, desired: <String, dynamic>{'browserName': packageTestRuntime.identifier});
+        return WebDriverBrowser(driver, url);
+      } on SocketException {
+        // Sometimes we may try to connect before the web driver port is ready.
+        // So we should here. Note that if there was some issue with the
```
Comment (yjbanov) [inline lib/web_ui/dev/webdriver_browser.dart:54]: nit: missing word between "should" and "here"

File: lib/web_ui/dev/webdriver_browser.dart [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        // webdriver process, we may loop infinitely here, so we're relying on
+        // the test timeout to kill us if it takes too long to connect.
+        print('Failed to connect to webdriver process. Retrying in 100 ms');
+        await Future<void>.delayed(const Duration(milliseconds: 100));
+      } catch (exception) {
+        print('Exception while creating webdriver: $exception');
```
Comment (yjbanov) [inline lib/web_ui/dev/webdriver_browser.dart:60]: When rethrowing you don't need to also print, because the `rethrow` will retain all the information about the original error, including its stack trace.

File: lib/web_ui/dev/webdriver_browser.dart [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  @override
+  Future<Browser> launchBrowserInstance(Uri url, {bool debug = false}) async {
+    for (;;) {
+      try {
+        final WebDriver driver = await createDriver(
+          uri: driverUri, desired: <String, dynamic>{'browserName': packageTestRuntime.identifier});
+        return WebDriverBrowser(driver, url);
+      } on SocketException {
```
Comment (yjbanov) [inline lib/web_ui/dev/webdriver_browser.dart:52]: I wonder if it would be more appropriate to smoke test (and retry if necessary) the driver process in the `prepare` method. This way other methods that need the driver to be ready don't need to mistrust the initialization process of the driver.

File: lib/web_ui/dev/webdriver_browser.dart [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import 'browser.dart';
+
+abstract class WebDriverBrowserEnvironment extends BrowserEnvironment {
+  late final Process _driverProcess;
+
+  Future<Process> spawnDriverProcess();
+  Uri get driverUri;
+
+  @override
+  Future<void> prepare() async {
```
Comment (yjbanov) [inline lib/web_ui/dev/webdriver_browser.dart:22]: Sometimes there's an existing driver process hanging around on the system. It may be useful to check that the port is not already taken. Otherwise, the test might end up using the wrong driver process (depending on when exactly the newly spawned process crashes due to the taken port).

File: lib/web_ui/dev/webdriver_browser.dart [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        await Future<void>.delayed(const Duration(milliseconds: 100));
+      } catch (exception) {
+        print('Exception while creating webdriver: $exception');
+        rethrow;
+      }
+    }
+  }
+}
+
+class WebDriverBrowser extends Browser {
```
Comment (yjbanov) [inline lib/web_ui/dev/webdriver_browser.dart:67]: I _love_ how simple the driver-based implementation is!

File: lib/web_ui/dev/webdriver_browser.dart (hunk 7 above)
Comment (eyebrowsoffire) [reply lib/web_ui/dev/webdriver_browser.dart:22]: I found a nice simple implementation of a "find unused port" function in the dart SDK, which I am shamefully stealing here. That should solve this problem.

File: lib/web_ui/dev/webdriver_browser.dart (hunk 6 above)
Comment (eyebrowsoffire) [reply lib/web_ui/dev/webdriver_browser.dart:52]: I tried coming up with a good solution here based on your suggestion, but I'm not really sure I really like anything I could come up with. I understand where you're coming from in that it seems like the responsibility of "prepare" to ensure that the web driver process is actually ready to go, but I think actually trying to enforce this only ends up with negatives practically speaking. For one thing, doing the smoke test in `prepare` results in opening an extra socket just to check if the driver process is ready, which is a non-trivial operation from a perf perspective. In practice, this is being validated the first time `launchBrowserInstance` is invoked, and after that, other browser launches should end up in the happy path. The only real benefit I could see to doing the validation in prepare is that if there is some situation where the web driver launches and the smoke test passes, but something bad happens in the web driver process before a subsequent invocation of `launchBrowserInstance` then we would fail fast rather than have to wait for the test timeout. But this seems like an edge case to optimize for that (hopefully) wouldn't happen too often in practice anyway.
